src/pykeen/pipeline/api.py

```diff
@@ -1003,11 +1012,46 @@ def pipeline(  # noqa: C901
     logging.debug("Evaluation will be run with following parameters:")
     logging.debug(f"evaluation_kwargs: {evaluation_kwargs}")
     evaluate_start_time = time.time()
-    metric_results: MetricResults = evaluator_instance.evaluate(
-        model=model_instance,
-        mapped_triples=mapped_triples,
-        **evaluation_kwargs,
-    )
+
+    # Begin an infinite loop to go through several possible training failure recovery scenarios
+    while True:
```

Comment:
maybe this would be easier to understand if you broke it out into a function that used recursion instead of infinite looping. The possible tree that it can follow isn't so deep, so maybe infinite loop isn't the best way to write it?

Reply:
I don't think that recursion makes sense here, since the logic remains case specific and therefore we would still need the case distinctions. For the same reason it is an infinite loop, because we don't know the current state and the entry point to the case tree. Anticipating this would only make the code more complicated Imo. Do you have a concrete suggestion?